rpcs3/Emu/Audio/FAudio/FAudioBackend.cpp

```diff
@@ -49,127 +62,192 @@ FAudioBackend::~FAudioBackend()
-		return false;
+		FAudio_.error("FAudioSourceVoice_SetFrequencyRatio() failed(0x%08x)", res);
+		return 1.0f;
 	}
 
-	return true;
+	return new_ratio;
+}
+
+void FAudioBackend::SetWriteCallback(std::function<u32(u32, void *)> cb)
+{
+	std::lock_guard lock(m_cb_mutex);
+	m_write_callback = cb;
 }
 
-u64 FAudioBackend::GetNumEnqueuedSamples()
+f64 FAudioBackend::GetCallbackFrameLen()
 {
-	FAudioVoiceState state;
-	FAudioSourceVoice_GetState(m_source_voice, &state, 0);
+	ensure(m_source_voice != nullptr);
+
+	f64 min_latency{};
+	const f64 _10ms = 0.01;
 
-	// all buffers contain AUDIO_BUFFER_SAMPLES, so we can easily calculate how many samples there are remaining
-	return (AUDIO_BUFFER_SAMPLES - state.SamplesPlayed % AUDIO_BUFFER_SAMPLES) + (state.BuffersQueued * AUDIO_BUFFER_SAMPLES);
+	u32 samples_per_q = 0, freq = 0;
+	FAudio_GetProcessingQuantum(m_instance, &samples_per_q, &freq);
+
+	if (freq)
+	{
+		min_latency = static_cast<f64>(samples_per_q) / freq;
+	}
+
+	return std::max<f64>(min_latency, _10ms);
 }
 
-f32 FAudioBackend::SetFrequencyRatio(f32 new_ratio)
+void FAudioBackend::OnVoiceProcessingPassStart_func(FAudioVoiceCallback *cb_obj, u32 BytesRequired)
 {
-	new_ratio = std::clamp(new_ratio, FAUDIO_MIN_FREQ_RATIO, FAUDIO_DEFAULT_FREQ_RATIO);
+	FAudioBackend *faudio = static_cast<FAudioBackend *>(cb_obj);
 
-	const u32 res = FAudioSourceVoice_SetFrequencyRatio(m_source_voice, new_ratio, FAUDIO_COMMIT_NOW);
-	if (res)
+	std::unique_lock lock(faudio->m_cb_mutex, std::defer_lock);
+	if (BytesRequired && lock.try_lock() && faudio->m_write_callback && faudio->m_playing)
 	{
-		FAudio_.fatal("FAudioSourceVoice_SetFrequencyRatio() failed(0x%08x)", res);
-		return 1.0f;
+		ensure(BytesRequired <= faudio->m_data_buf_len, "FAudio internal buffer is too small. Report to developers!");
+
+		const u32 sample_size = faudio->get_sample_size() * faudio->get_channels();
+		u32 written = std::min(faudio->m_write_callback(BytesRequired, faudio->m_data_buf.get()), BytesRequired);
+		written -= written % sample_size;
+
+		if (written >= sample_size)
+		{
+			memcpy(faudio->m_last_sample, faudio->m_data_buf.get() + written - sample_size, sample_size);
+		}
+
+		for (u32 i = written; i < BytesRequired; i += sample_size)
+		{
+			memcpy(faudio->m_data_buf.get() + i, faudio->m_last_sample, sample_size);
+		}
+
+		FAudioBuffer buffer{};
+		buffer.AudioBytes = BytesRequired;
+		buffer.LoopBegin  = FAUDIO_NO_LOOP_REGION;
+		buffer.pAudioData = static_cast<const u8*>(faudio->m_data_buf.get());
+
+		const u32 res = FAudioSourceVoice_SubmitSourceBuffer(faudio->m_source_voice, &buffer, nullptr);
+		if (res)
+		{
+			FAudio_.error("FAudioSourceVoice_SubmitSourceBuffer() failed(0x%08x)", res);
+		}
 	}
+}
 
-	return new_ratio;
+void FAudioBackend::OnCriticalError_func(FAudioEngineCallback *cb_obj, u32 Error)
+{
+	FAudioBackend *faudio = static_cast<FAudioBackend *>(cb_obj);
+
+	FAudio_.error("OnCriticalError() failed(0x%08x)", Error);
```

Comment:
the error should always be logged first, in case the rest of the code fails for whatever reason